In [1]:
# Импорт всех необходимых библиотек
!pip install catboost xgboost

import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, f1_score, roc_auc_score
from catboost import CatBoostClassifier
import xgboost as xgb

data = pd.read_excel('Данные_для_курсовои_Классическое_МО.xlsx').drop(columns=['Unnamed: 0'])
data.dropna(inplace=True)

# Переместим названия колонок в отдельную переменную, пригодится
columns = data.columns.tolist()

# Быстрая проверка
data.head()

#'IC50, mM', 'CC50, mM', 'SI'

IC50, mM    CC50, mM         SI  MaxAbsEStateIndex  MaxEStateIndex  \
0    6.239374  175.482382  28.125000           5.094096        5.094096   
1    0.771831    5.402819   7.000000           3.961417        3.961417   
2  223.808778  161.142320   0.720000           2.627117        2.627117   
3    1.705624  107.855654  63.235294           5.097360        5.097360   
4  107.131532  139.270991   1.300000           5.150510        5.150510   

   MinAbsEStateIndex  MinEStateIndex       qed        SPS    MolWt  ...  \
0           0.387225        0.387225  0.417362  42.928571  384.652  ...   
1           0.533868        0.533868  0.462473  45.214286  388.684  ...   
2           0.543231        0.543231  0.260923  42.187500  446.808  ...   
3           0.390603        0.390603  0.377846  41.862069  398.679  ...   
4           0.270476        0.270476  0.429038  36.514286  466.713  ...   

   fr_sulfide  fr_sulfonamd  fr_sulfone  fr_term_acetylene  fr_tetrazole  \
0           0             0           0                  0             0   
1           0             0           0                  0             0   
2           0             0           0                  0             0   
3           0             0           0                  0             0   
4           0             0           0                  0             0   

   fr_thiazole  fr_thiocyan  fr_thiophene  fr_unbrch_alkane  fr_urea  
0            0            0             0                 3        0  
1            0            0             0                 3        0  
2            0            0             0                 3        0  
3            0            0             0                 4        0  
4            0            0             0                 0        0  

[5 rows x 213 columns]

In [2]:
# Создаём бинарную целевую переменную
si_median = data['SI'].median()
print(f"Медиана SI: {si_median}")

y = (data['SI'] > si_median).astype(int)
X = data.drop(columns=['IC50, mM', 'CC50, mM', 'SI'])

# Разбиение на обучающую и тестовую выборку
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)

# Масштабирование признаков
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

Медиана SI: 3.856410256410256


Логистическая регрессия
---

In [3]:
logreg = LogisticRegression(max_iter=1000, random_state=42)
logreg.fit(X_train_scaled, y_train)
y_pred_logreg = logreg.predict(X_test_scaled)

print("Логистическая регрессия:")
print("Accuracy:", accuracy_score(y_test, y_pred_logreg))
print("F1:", f1_score(y_test, y_pred_logreg))
print("ROC-AUC:", roc_auc_score(y_test, y_pred_logreg))

Логистическая регрессия:
Accuracy: 0.62
F1: 0.6161616161616161
ROC-AUC: 0.62


In [4]:
rf_model = RandomForestClassifier(n_estimators=200, max_depth=7, random_state=42)
rf_model.fit(X_train, y_train)
y_pred_rf = rf_model.predict(X_test)

print("Random Forest:")
print("Accuracy:", accuracy_score(y_test, y_pred_rf))
print("F1:", f1_score(y_test, y_pred_rf))
print("ROC-AUC:", roc_auc_score(y_test, y_pred_rf))

Random Forest:
Accuracy: 0.66
F1: 0.6304347826086957
ROC-AUC: 0.66


In [5]:
cat_model = CatBoostClassifier(
    iterations=1000,
    depth=6,
    learning_rate=0.05,
    random_seed=42,
    verbose=0
)
cat_model.fit(X_train, y_train)
y_pred_cat = cat_model.predict(X_test)

print("CatBoost:")
print("Accuracy:", accuracy_score(y_test, y_pred_cat))
print("F1:", f1_score(y_test, y_pred_cat))
print("ROC-AUC:", roc_auc_score(y_test, y_pred_cat))

CatBoost:
Accuracy: 0.62
F1: 0.62
ROC-AUC: 0.62


In [6]:
xgb_model = xgb.XGBClassifier(
    n_estimators=500,
    max_depth=5,
    learning_rate=0.05,
    random_state=42
)
xgb_model.fit(X_train, y_train)
y_pred_xgb = xgb_model.predict(X_test)

print("XGBoost:")
print("Accuracy:", accuracy_score(y_test, y_pred_xgb))
print("F1:", f1_score(y_test, y_pred_xgb))
print("ROC-AUC:", roc_auc_score(y_test, y_pred_xgb))

XGBoost:
Accuracy: 0.62
F1: 0.6122448979591837
ROC-AUC: 0.62


In [7]:
results_si_clf = pd.DataFrame({
    'Модель': ['Logistic Regression', 'Random Forest', 'CatBoost', 'XGBoost'],
    'Accuracy': [
        accuracy_score(y_test, y_pred_logreg),
        accuracy_score(y_test, y_pred_rf),
        accuracy_score(y_test, y_pred_cat),
        accuracy_score(y_test, y_pred_xgb)
    ],
    'F1': [
        f1_score(y_test, y_pred_logreg),
        f1_score(y_test, y_pred_rf),
        f1_score(y_test, y_pred_cat),
        f1_score(y_test, y_pred_xgb)
    ],
    'ROC-AUC': [
        roc_auc_score(y_test, y_pred_logreg),
        roc_auc_score(y_test, y_pred_rf),
        roc_auc_score(y_test, y_pred_cat),
        roc_auc_score(y_test, y_pred_xgb)
    ]
})

display(results_si_clf)

Модель  Accuracy        F1  ROC-AUC
0  Logistic Regression      0.62  0.616162     0.62
1        Random Forest      0.66  0.630435     0.66
2             CatBoost      0.62  0.620000     0.62
3              XGBoost      0.62  0.612245     0.62

Все модели показывают очень схожие и скромные результаты.

Разброс точности: от 62% до 66%, что говорит о достаточно слабой предсказуемости признаков в контексте SI.

CatBoost и XGBoost — ожидались как более мощные, но по факту не дали прироста.

Возможно, данные не содержат глубокой нелинейности или просто слабый сигнал.

---
Random Forest немного выделился, показав лучшие метрики, особенно по точности. Он мог зацепиться за какие-то закономерности, недоступные логистике.

---
Logistic Regression — как базовая модель, справилась на 62%

Можно попробовать:

- провести дополнительный feature engineering ( логарифмировать SI или признаки)


- использовать мультимодельный ансамбль или трансформеры.

